### PR TITLE
Add governance guardrails tests for update pparams

### PR DESCRIPTION
**Pull request by nabinpkl**
This PR proposes tests for governance guardrails. This is based on the Plutus Governance guardrails script and default constitution. At first, the script is enacted and then checked for unconstitutional values/out-of-range values for all protocol parameters defined in the constitution along with a check to ensure that valid values are accepted as a protocol parameter governance action.

The parameters and their values are checked based on Plutus cardano-constitution.

commit [db119872c0dfc8537a39d38cea8ef0b20c59ecb5](https://github.com/IntersectMBO/plutus/tree/db119872c0dfc8537a39d38cea8ef0b20c59ecb5/cardano-constitution)

The default constitution file
https://github.com/IntersectMBO/plutus/blob/db119872c0dfc8537a39d38cea8ef0b20c59ecb5/cardano-constitution/data/defaultConstitution.json

The user stories are derived from this PR.
https://github.com/IntersectMBO/cardano-test-plans/pull/52

This PR is ready for review and waiting for user stories to be merged.

### CODE REVIEW

**Review by mkoura — CHANGES_REQUESTED** [review]
Thanks for the PR. So far I took only brief look and have several comments.

There can be improvements made to enable parallel run of the tests (they don't change the environment, apart from the setup, so I believe they can run in parallel) and collateral handling.

If you don't mind, I'll do the changes, it will be faster for me than writing detailed comments.

File: cardano_node_tests/tests/common.py [hunk 1]
```diff
@@ -322,3 +322,38 @@ def get_conway_address_deposit(cluster_obj: clusterlib.ClusterLib) -> int:
         stake_deposit_amt = cluster_obj.g_query.get_address_deposit()
 
     return stake_deposit_amt
+
+
+def create_collaterals(
```
[the remark this answers is not in the document]
Comment (nabinpkl) [reply cardano_node_tests/tests/common.py]: Thank you for the feedback. This has been moved to clustelib_utils with the slicing of utxos on the latest update.

File: cardano_node_tests/tests/common.py [hunk 2]
```diff
@@ -322,3 +322,38 @@ def get_conway_address_deposit(cluster_obj: clusterlib.ClusterLib) -> int:
...
+    )
+    cluster.g_transaction.submit_tx(tx_file=tx_signed, txins=tx_output.txins)
+
+    utxos = cluster.g_query.get_utxo(tx_raw_output=tx_output)
+    utxo_ix_offset = clusterlib_utils.get_utxo_ix_offset(utxos=utxos, txouts=tx_output.txouts)
+    # Return collateral UTxOs according to the order in `tx_outs`
+    return [
+        clusterlib.filter_utxos(utxos=utxos, utxo_ix=utxo_ix_offset + i)
+        for i in range(len(tx_outs))
+    ]
```
[the remark this answers is not in the document]
Comment (nabinpkl) [reply cardano_node_tests/tests/common.py]: This has been moved to clustelib_utils with the slicing of utxos on the latest update.

File: cardano_node_tests/tests/tests_conway/test_guardrails.py [hunk 3]
```diff
@@ -0,0 +1,1543 @@
...
+    return conway_common.get_registered_pool_user(
+        cluster_manager=cluster_manager,
+        name_template=name_template,
+        cluster_obj=cluster,
+        caching_key=key,
+        fund_amount=100000_000_000,
+    )
+
+
+class ClusterWithConstitutionRecord:
```
[the remark this answers is not in the document]
Comment (nabinpkl) [reply cardano_node_tests/tests/tests_conway/test_guardrails.py:80]: Right. This has been updated to simpler dataclass.

File: cardano_node_tests/tests/tests_conway/test_guardrails.py [hunk 4]
```diff
@@ -0,0 +1,1543 @@
...
+
+        constitution_url = "http://www.const-with-plutus.com"
+        constitution_hash = "0000000000000000000000000000000000000000000000000000000000000000"
+
+        deposit_amt = cluster.conway_genesis["govActionDeposit"]
+        prev_action_rec = governance_utils.get_prev_action(
+            action_type=governance_utils.PrevGovActionIds.CONSTITUTION,
+            gov_state=cluster.g_conway_governance.query.gov_state(),
+        )
+        constitution_action = cluster.g_conway_governance.action.create_constitution(
```
[the remark this answers is not in the document]
Comment (nabinpkl) [reply cardano_node_tests/tests/tests_conway/test_guardrails.py]: `propose_change_constitution` has been updated with additional optional `constitution_script_hash` parameter.

File: cardano_node_tests/tests/tests_conway/test_guardrails.py [hunk 5]
```diff
@@ -0,0 +1,1543 @@
...
+        deposit_amt=deposit_amt,
+        anchor_url=anchor_url,
+        anchor_data_hash=anchor_data_hash,
+        cli_args=update_args,
+        prev_action_txid=prev_action_rec.txid,
+        prev_action_ix=prev_action_rec.ix,
+        deposit_return_stake_vkey_file=pool_user_lg.stake.vkey_file,
+    )
+
+    # Create collateral utxo for plutus script
```
[the remark this answers is not in the document]
Comment (nabinpkl) [reply cardano_node_tests/tests/tests_conway/test_guardrails.py]: The collateral creation step has been moved to set up.

File: cardano_node_tests/tests/tests_conway/test_guardrails.py [hunk 6]
```diff
@@ -0,0 +1,1543 @@
...
+    param_name: str  # name of the protocol parameter
+    param_lower_limit: int | None = None  # optional lower limit of the parameter
+    param_upper_limit: int | None = None  # optional upper limit of the parameter
+
+
+def check_min_value_proposals(
+    cluster_with_constitution: ClusterWithConstitutionRecord,
+    param: GuardrailTestParam,
+    min_value: int | float,
+    dependent_proposals: List[clusterlib_utils.UpdateProposal] | Tuple,
```
[the remark this answers is not in the document]
Comment (nabinpkl) [reply cardano_node_tests/tests/tests_conway/test_guardrails.py]: Right. It has been updated to use Union instead.

File: cardano_node_tests/tests/tests_conway/test_guardrails.py [hunk 7]
```diff
@@ -0,0 +1,1543 @@
...
+        assert new_constitution["script"] == constitution_script_hash
+
+    cur_constitution = cluster.g_conway_governance.query.constitution()
+
+    # Enact the new constitution if the current one is not the one we expect
+    if cur_constitution.get("script") != constitution_script_hash:
+        _enact_script_constitution()
+
+    # Testnet will be using an unexpected constitution, respin is needed
+    cluster_manager.set_needs_respin()
```
[the remark this answers is not in the document]
Comment (nabinpkl) [reply cardano_node_tests/tests/tests_conway/test_guardrails.py]: I am not familiar much with the framework you mentioned. This step has been removed and I believe it needs to be updated in the CI/CD pipeline for running guardrails test.

File: cardano_node_tests/tests/tests_conway/test_guardrails.py [hunk 8]
```diff
@@ -0,0 +1,1543 @@
...
+    return action_txid
+
+
+def check_invalid_proposals(
+    cluster_with_constitution: ClusterWithConstitutionRecord,
+    proposals: List[clusterlib_utils.UpdateProposal],
+):
+    """Check that the guardrails are enforced."""
+    with pytest.raises(clusterlib.CLIError) as excinfo:
+        propose_param_changes(cluster_with_constitution, proposals)
```
[the remark this answers is not in the document]
Comment (nabinpkl) [reply cardano_node_tests/tests/tests_conway/test_guardrails.py]: All the function calls with more than one argument are updated to use kwargs.

**Comment by nabinpkl** [comment]
> Thanks for the PR. So far I took only brief look and have several comments.
> 
> There can be improvements made to enable parallel run of the tests (they don't change the environment, apart from the setup, so I believe they can run in parallel) and collateral handling.
> 
> If you don't mind, I'll do the changes, it will be faster for me than writing detailed comments.

Thank you for the review comments and feedback.
I have addressed some of the comments.

We'd welcome the changes you'd like to make to the PR.

It'd be great if you could push the changes directly on the PR, get it merged if possible, or merge PR and then make the necessary changes to run them in parallel. 

Or I can address those changes If there are some examples of tests currently running in parallel and how the setup is being done.

**Comment by mkoura** [comment]
Looks like I cannot push to your PR. Can you give me access?

**Comment by nabinpkl** [comment]
> Looks like I cannot push to your PR. Can you give me access?

I just sent you an invite. It should give access now.

**Review by mkoura — APPROVED** [review]

**Pull request merged**